src/mon/MgrStatMonitor.cc

```diff
@@ -66,6 +66,79 @@ void MgrStatMonitor::create_initial()
   encode(service_map, pending_service_map_bl, CEPH_FEATURES_ALL);
 }
 
+void MgrStatMonitor::calc_pool_availability()
+{
+  dout(20) << __func__ << dendl;
+  auto pool_avail_end = pool_availability.end();
+  for (const auto& i : digest.pool_pg_unavailable_map) {
+    const auto& poolid = i.first;
+    if (pool_availability.find(poolid) == pool_avail_end){
+      // New Pool so we add.
+      pool_availability.insert({poolid, PoolAvailability()});
+      dout(20) << __func__ << "Adding pool: " << poolid << dendl;
+    }
+  }
+  utime_t now(ceph_clock_now());
+  auto pool_unavail_end = digest.pool_pg_unavailable_map.end();
+  for (const auto& i : pool_availability) {
+    const auto& poolid = i.first;
+    if (digest.pool_pg_unavailable_map.find(poolid) ==
+      pool_unavail_end) {
+      // delete none exist pool
+      pool_availability.erase(poolid);
+      dout(20) << __func__ << "Deleting pool: " << poolid << dendl;
+      continue;
+    }
+    if (mon.osdmon()->osdmap.have_pg_pool(poolid)){
+      // Currently, couldn't find an elegant way to get pool name
+      pool_availability[poolid].pool_name = mon.osdmon()->osdmap.get_pool_name(poolid);
```

Comment:
Are we sure that digest and mon.osdmon()->osdmap are from the same epoch?  Can digest include pools not present in mon.osdmon()->osdmap?  If you aren't sure, guard against mon.osdmon()->osdmap not containing poolid (perhaps by using "<unknown>" as the pool) to ensure that get_pool_name doesn't assert.